FWCore/Framework/src/EventProcessor.cc

```diff
@@ -718,113 +716,148 @@ namespace edm {
+      }
+    } catch (...) {
+      if (!firstException) {
+        firstException = std::current_exception();
+      }
+    }
+    for (auto& subProcess : subProcesses_) {
+      CMS_SA_ALLOW try { subProcess.doBeginJob(); } catch (...) {
+        if (!firstException) {
+          firstException = std::current_exception();
+        }
+      }
+    }
+    if (firstException) {
+      std::rethrow_exception(firstException);
+    }
+
+    beginJobSucceeded_ = true;
+    beginStream();
+  }
 
+  void EventProcessor::beginStream() {
+    // This will process streams concurrently, but not modules in the
+    // same stream or SubProcesses.
     oneapi::tbb::task_group group;
-    FinalWaitingTask last{group};
+    FinalWaitingTask finalWaitingTask{group};
     using namespace edm::waiting_task::chain;
-    first([this](auto nextTask) {
+    {
+      WaitingTaskHolder taskHolder(group, &finalWaitingTask);
       for (unsigned int i = 0; i < preallocations_.numberOfStreams(); ++i) {
-        first([i, this](auto nextTask) {
-          ServiceRegistry::Operate operate(serviceToken_);
-          schedule_->beginStream(i);
-        }) | ifThen(not subProcesses_.empty(), [this, i](auto nextTask) {
-          ServiceRegistry::Operate operate(serviceToken_);
-          for_all(subProcesses_, [i](auto& subProcess) { subProcess.doBeginStream(i); });
-        }) | lastTask(nextTask);
+        first([this, i](auto nextTask) {
+          std::exception_ptr exceptionPtr;
+          {
+            ServiceRegistry::Operate operate(serviceToken_);
+            CMS_SA_ALLOW try { schedule_->beginStream(i); } catch (...) {
+              exceptionPtr = std::current_exception();
+            }
+            for (auto& subProcess : subProcesses_) {
+              CMS_SA_ALLOW try { subProcess.doBeginStream(i); } catch (...) {
+                if (!exceptionPtr) {
+                  exceptionPtr = std::current_exception();
+                }
+              }
+            }
+          }
+          nextTask.doneWaiting(exceptionPtr);
+        }) | lastTask(taskHolder);
       }
-    }) | runLast(WaitingTaskHolder(group, &last));
-    last.wait();
+    }
+    finalWaitingTask.wait();
   }
 
   void EventProcessor::endJob() {
     // Collects exceptions, so we don't throw before all operations are performed.
     ExceptionCollector c(
-        "Multiple exceptions were thrown while executing endJob. An exception message follows for each.\n");
+        "Multiple exceptions were thrown while executing endStream and endJob. An exception message follows for "
+        "each.\n");
 
     //make the services available
     ServiceRegistry::Operate operate(serviceToken_);
 
-    using namespace edm::waiting_task::chain;
+    if (beginJobSucceeded_) {
+      endStream(c);
+    }
 
-    oneapi::tbb::task_group group;
-    edm::FinalWaitingTask waitTask{group};
+    if (beginJobStartedModules_) {
+      try {
+        convertException::wrap([this, &c]() { schedule_->endJob(c); });
```

Comment:
Maybe we should mark `ExceptionCollector::addException()` (and `call()`) as `noexcept`. While strictly speaking both functions do operations that may throw exceptions, but if that happens the job is probably in so bad state anyway that terminating the process immediately could be a reasonable action.

Reply:
I was worried about that. I already declared Schedule::endStream and SubProcess::doEndStream as noexcept. ExceptionCollector emitting an exception that propagated out via those into EventProcessor::endStream (soon to be endStreams) would cause a problem. In all the other places ExceptionCollector is currently used, it looks like the exception would propagate all the way out to the top levels of cmsRun and not cause a problem. I think that an exception would be better than calling terminate in the other cases. Although, if we add more concurrency into endJob and endStream that might change...